File: tests/test_overload.py
import unittest
import sys
import jep
from java.util import ArrayList
TestOverload = jep.findClass('jep.test.TestOverload')


class TestTypes(unittest.TestCase):

    def test_any_primitive(self):
        if sys.version_info.major == 2:
            self.assertEqual(TestOverload.any_primitive(0), 'int')
            self.assertEqual(TestOverload.any_primitive(1), 'int')
        else:
            self.assertEqual(TestOverload.any_primitive(0), 'long')
            self.assertEqual(TestOverload.any_primitive(1), 'long')
        # float and double methods are both ranked the same so it is arbitrary
        # which one you get, this may be an area for improvement
        self.assertIn(TestOverload.any_primitive(0.0), ('float', 'double'))
        self.assertEqual(TestOverload.any_primitive(True), 'boolean')

    def test_boxing(self):
        self.assertEqual(TestOverload.int_or_Integer(0), 'int')
        self.assertEqual(TestOverload.int_or_Integer(1), 'int')
        self.assertEqual(TestOverload.int_or_Integer(None), 'Integer')
        self.assertEqual(TestOverload.int_or_Object(0), 'int')
        self.assertEqual(TestOverload.int_or_Object(1), 'int')
        self.assertEqual(TestOverload.int_or_Object(None), 'Object')
        self.assertEqual(TestOverload.int_or_Object('0'), 'Object')
        self.assertEqual(TestOverload.float_or_Float(0.0), 'float')
        self.assertEqual(TestOverload.float_or_Float(1.1), 'float')

    def test_string(self):
        # Not sure this one is actually what we want?
        self.assertEqual(TestOverload.char_or_String('a'), 'String')
        self.assertEqual(TestOverload.char_or_String('abc'), 'String')
        self.assertEqual(TestOverload.char_or_String(None), 'String')
        self.assertEqual(TestOverload.Object_or_String('a'), 'String')
        self.assertEqual(TestOverload.Object_or_String('abc'), 'String')
        self.assertEqual(TestOverload.Object_or_String(None), 'Object')
        self.assertEqual(TestOverload.Object_or_String(0), 'Object')
        self.assertEqual(TestOverload.Object_or_String(0.0), 'Object')

    def test_collections(self):
        self.assertEqual(TestOverload.Object_or_ArrayList(ArrayList()), 'ArrayList')
        self.assertEqual(TestOverload.ArrayList_or_List(ArrayList()), 'ArrayList')
        self.assertEqual(TestOverload.Object_or_Array(ArrayList().toArray()), 'Array')

    def test_class(self):
        self.assertEqual(TestOverload.Object_or_Class(ArrayList), 'Class')
        self.assertEqual(TestOverload.Object_or_Class(0), 'Object')
        self.assertEqual(TestOverload.Object_or_Class(0.0), 'Object')
        self.assertEqual(TestOverload.Object_or_Class('0'), 'Object')
        # This probably doesn't matter but it is currently deterministic
        self.assertEqual(TestOverload.Object_or_Class(None), 'Object')

    def things_that_might_need_fixing(self):
        # 64 bit python 2 stores this in an int but it is too big for a java int
        # The method matching doesn't actually check the size of the int so it
        # chooses the int method and gets a TypeError
        self.assertEqual(TestOverload.any_primitive(2147483648), 'long')
        # The conversion code will treat a tuple as a boolean but the method
        # matching code does not.
        self.assertEqual(TestOverload.any_primitive(()), 'boolean')
        # Matching does not allow ints to be floats.
        self.assertEqual(TestOverload.float_or_Float(0), 'float')
        self.assertEqual(TestOverload.float_or_Float(1), 'float')
        # All instances of Object are matched equally
        self.assertEqual(TestOverload.Object_or_Integer(0), 'Integer')
        # Method choosing doesn't know about unicode so picks an arbitrary one
        self.assertEqual(TestOverload.char_or_String(u'\u263A'), 'char')
        self.assertEqual(TestOverload.char_or_String(u'\U0001F604'), 'String')
        # The method matching doesn't handle lists so it picks an arbitrary method
        self.assertEqual(TestOverload.Object_or_List([]), 'List')
        self.assertEqual(TestOverload.Object_or_ArrayList([]), 'ArrayList')
        self.assertEqual(TestOverload.ArrayList_or_List([]), 'ArrayList')
        # The method matching doesn't handle tuples so it picks an arbitrary method
        self.assertEqual(TestOverload.Object_or_List(()), 'List')
        self.assertEqual(TestOverload.Object_or_ArrayList(()), 'Object')
        self.assertEqual(TestOverload.ArrayList_or_List(()), 'List')
        # interfaces are not prefered over Object so its arbitrary
        self.assertEqual(TestOverload.Object_or_List(ArrayList()), 'List')
        # The method matching doesn't handle dicts so it picks an arbitrart method
        self.assertEqual(TestOverload.Object_or_Map({}), 'Map')
        # The method choosing treats all Objects as equal and picks an arbitrary one
        self.assertEqual(TestOverload.Object_or_List(0), 'Object')
        self.assertEqual(TestOverload.Object_or_List(0.0), 'Object')
        self.assertEqual(TestOverload.Object_or_ArrayList(0), 'Object')
        self.assertEqual(TestOverload.Object_or_ArrayList(0.0), 'Object')
        self.assertEqual(TestOverload.Object_or_Map(0), 'Object')
        self.assertEqual(TestOverload.Object_or_Map(0.0), 'Object')
        self.assertEqual(TestOverload.Object_or_Array(0), 'Object')
        self.assertEqual(TestOverload.Object_or_Array(0.0), 'Object')

File: src/jep/test/TestOverload.java
package jep.test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Created: December 2016
 * 
 * @author Ben Steffensmeier
 * @since 3.7
 */
public class TestOverload {

    public static String any_primitive(long l){
        return "long";
    }

    public static String any_primitive(int i){
        return "int";
    }

    public static String any_primitive(short s){
        return "short";
    }

    public static String any_primitive(byte b){
        return "byte";
    }

    public static String any_primitive(char c){
        return "char";
    }

    public static String any_primitive(boolean b){
        return "boolean";
    }

    public static String any_primitive(float f){
        return "float";
    }

    public static String any_primitive(double d){
        return "double";
    }

    public static String int_or_Integer(int i){
        return "int";
    }

    public static String int_or_Integer(Integer i){
        return "Integer";
    }

    public static String int_or_Object(int i){
        return "int";
    }

    public static String int_or_Object(Object o){
        return "Object";
    }

    public static String Object_or_Integer(Object o){
        return "Object";
    }

    public static String Object_or_Integer(Integer i){
        return "Integer";
    }

    public static String float_or_Float(float f){
        return "float";
    }

    public static String float_or_Float(Float f){
        return "Float";
    }

    public static String char_or_String(char c){
        return "char";
    }

    public static String char_or_String(String s){
        return "String";
    }

    public static String Object_or_String(Object o){
        return "Object";
    }

    public static String Object_or_String(String s){
        return "String";
    }

    public static String Object_or_List(Object o){
        return "Object";
    }

    public static String Object_or_List(List l){
        return "List";
    }

    public static String Object_or_ArrayList(Object o){
        return "Object";
    }

    public static String Object_or_ArrayList(ArrayList a){
        return "ArrayList";
    }

    public static String ArrayList_or_List(ArrayList a){
        return "ArrayList";
    }

    public static String ArrayList_or_List(List l){
        return "List";
    }

    public static String Object_or_Map(Object o){
        return "Object";
    }

    public static String Object_or_Map(Map m){
        return "Map";
    }

    public static String Object_or_Array(Object o){
        return "Object";
    }

    public static String Object_or_Array(Object[] a){
        return "Array";
    }

    public static String Object_or_Class(Object o){
        return "Object";
    }

    public static String Object_or_Class(Class<?> c){
        return "Class";
    }
}
